tasks/pdp/proofset_addroot_watch.go

```diff
@@ -111,6 +111,30 @@ func processDataSetPieceAdd(ctx context.Context, db *harmonydb.DB, ethClient *et
 }
 
 func extractAndInsertPiecesFromReceipt(ctx context.Context, db *harmonydb.DB, receipt *types.Receipt, pieceAdd DataSetPieceAdd) error {
+	resolvedDataSetId := pieceAdd.DataSet
+	if !resolvedDataSetId.Valid {
+		var err error
+		resolvedDataSetId.Int64, err = extractDataSetIdFromReceipt(receipt)
+		if err != nil {
+			return fmt.Errorf("expeted to find dataSetId in receipt but failed to extract: %w", err)
+		}
+		resolvedDataSetId.Valid = true
+		var exists bool
+		// we check if the dataset exists already to avoid foreign key violation
+		err = db.QueryRow(ctx, `
+			SELECT EXISTS (
+				SELECT 1
+				FROM pdp_data_sets
+				WHERE id = $1
+			)`, resolvedDataSetId.Int64).Scan(&exists)
+		if err != nil {
+			return fmt.Errorf("failed to check if data set exists: %w", err)
+		}
+		if !exists {
+			// XXX: maybe return nil instead to avoid warning?
+			return fmt.Errorf("data set %d not found in pdp_data_sets", resolvedDataSetId.Int64)
+		}
+	}
```

Comment:
I would prefer not try to create_watch here as well. Just scan from DB and if it is NULL then just skip processing addPiece for this tipset. But that might cause unnecessary delay of 1 tipset. Any other ideas do this this without 2 watches?

Reply:
> Just scan from DB and if it is NULL then just skip processing addPiece for this tipset

That is what I'm doing here. If we get an add_piece triggered, and the dataset doesn't exist yet, we leave it alone and wait for another trigger for when the dataset is created.
A cleaner way might be to get `create_watch` to be processed with a higher priority than `add_pieces`. This could be achieved by combining them into one watcher and executing sequentially with create first.